### PR TITLE
JCLOUDS-903 - Support nova compute api extension server groups

### PR DESCRIPTION
**Pull request by VRanga000**

### CODE REVIEW

File: apis/openstack-nova/src/main/java/org/jclouds/openstack/nova/v2_0/domain/SchedulerHints.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+ * and <a href="http://docs.openstack.org/havana/config-reference/content/scheduler-filters.html#groupantiaffinityfilter"GroupAntiAffinity</a>
+ * Can be easily augmented to add support for other filters that can be configured via scheduler hints
+*/
+public class SchedulerHints {
+
+   public static Builder builder() {
+      return new ConcreteBuilder();
+   }
+
+   public Builder toBuilder() {
```
Comment (zack-shoylev) [inline apis/openstack-nova/src/main/java/org/jclouds/openstack/nova/v2_0/domain/SchedulerHints.java]: Do we want to move to the autovalue builders for this?

File: apis/openstack-nova/src/main/java/org/jclouds/openstack/nova/v2_0/domain/ServerGroup.java [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+   }
+
+   private static class ConcreteBuilder extends Builder<ConcreteBuilder> {
+      @Override
+      protected ConcreteBuilder self() {
+         return this;
+      }
+   }
+
+    private final String id;
```
Comment (zack-shoylev) [inline apis/openstack-nova/src/main/java/org/jclouds/openstack/nova/v2_0/domain/ServerGroup.java]: Formatting looks strange

File: apis/openstack-nova/src/main/java/org/jclouds/openstack/nova/v2_0/extensions/ServerGroupApi.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    * List all Security Groups.
+    *
+    * @return all Security Groups
+    */
+   @Named("ServerGroup:list")
+   @GET
+   @Path("/os-server-groups")
+   @SelectJson("server_groups")
+   @Fallback(EmptyFluentIterableOnNotFoundOr404.class)
+   FluentIterable<ServerGroup> list();
```
Comment (zack-shoylev) [inline apis/openstack-nova/src/main/java/org/jclouds/openstack/nova/v2_0/extensions/ServerGroupApi.java]: We should avoid using guava collections in user-facing interfaces. Also - does this need paging support?

File: apis/openstack-nova/src/test/java/org/jclouds/openstack/nova/v2_0/compute/NovaComputeServiceLiveTest.java [hunk 4]
```diff
@@ -36,7 +35,7 @@ public NovaComputeServiceLiveTest() {
       provider = "openstack-nova";
    }
 
-   @Override
+    @Override
```
Comment (zack-shoylev) [inline apis/openstack-nova/src/test/java/org/jclouds/openstack/nova/v2_0/compute/NovaComputeServiceLiveTest.java]: ?

File: apis/openstack-nova/src/test/java/org/jclouds/openstack/nova/v2_0/extensions/ServerGroupApiExpectTest.java [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+import static org.testng.Assert.assertFalse;
+import static org.testng.Assert.assertNull;
+import static org.testng.Assert.assertTrue;
+
+
+/**
+ * Tests ServerGroupApi guice wiring and parsing
+ */
+@Test(groups = "unit", testName = "ServerGroupApiExpectTest")
+public class ServerGroupApiExpectTest extends BaseNovaApiExpectTest {
```
Comment (zack-shoylev) [inline apis/openstack-nova/src/test/java/org/jclouds/openstack/nova/v2_0/extensions/ServerGroupApiExpectTest.java]: Is it feasible to switch to mock tests here in line with how new code is being implemented? Or do we want to keep expect tests for consistency.

File: apis/openstack-nova/src/test/java/org/jclouds/openstack/nova/v2_0/features/ServerApiLiveTest.java [hunk 6]
```diff
@@ -132,7 +134,57 @@ public void testCreateWithNetworkOptions() {
       }
    }
 
-   /**
+    /**
+     * This test creates a server with a scheduler hint that sets the server group
+     * used for <a href="http://docs.openstack.org/havana/config-reference/content/scheduler-filters.html#groupaffinityfilter"GroupAffinity</a>
```
Comment (zack-shoylev) [inline apis/openstack-nova/src/test/java/org/jclouds/openstack/nova/v2_0/features/ServerApiLiveTest.java]: I rather have the info directly in here in addition to the links

File: compute/src/test/java/org/jclouds/compute/internal/BaseComputeServiceLiveTest.java [hunk 7]
```diff
@@ -914,7 +912,7 @@ protected void tearDownContext() {
             // Destroy all nodes in the group but also make sure to destroy other created nodes that might not be in it.
             // The "testCreateTwoNodesWithOneSpecifiedName" creates nodes with an explicit name that puts them outside the group,
             // so the list of nodes should also be taken into account when destroying the nodes.
-            client.destroyNodesMatching(Predicates.<NodeMetadata> or(inGroup(group), in(nodes)));
+            computeService.destroyNodesMatching(Predicates.<NodeMetadata>or(inGroup(group), in(nodes)));
          }
       } catch (Exception e) {
```
Comment (zack-shoylev) [inline compute/src/test/java/org/jclouds/compute/internal/BaseComputeServiceLiveTest.java]: Thanks for the cleanup!

File: apis/openstack-nova/src/test/java/org/jclouds/openstack/nova/v2_0/extensions/ServerGroupApiExpectTest.java (hunk 5 above)
Comment (nacx) [inline apis/openstack-nova/src/test/java/org/jclouds/openstack/nova/v2_0/extensions/ServerGroupApiExpectTest.java]: We should be using mock tests in any new features we add, regardless of how existing tests are implemented.

File: apis/openstack-nova/src/main/java/org/jclouds/openstack/nova/v2_0/domain/SchedulerHints.java (hunk 1 above)
Comment (VRanga000) [inline apis/openstack-nova/src/main/java/org/jclouds/openstack/nova/v2_0/domain/SchedulerHints.java]: Thanks for the suggestion. Would you mind pointing me to an example?

File: apis/openstack-nova/src/main/java/org/jclouds/openstack/nova/v2_0/domain/ServerGroup.java (hunk 2 above)
Comment (VRanga000) [inline apis/openstack-nova/src/main/java/org/jclouds/openstack/nova/v2_0/domain/ServerGroup.java]: Will fix - odd that it got past checkstyle.

File: apis/openstack-nova/src/test/java/org/jclouds/openstack/nova/v2_0/extensions/ServerGroupApiExpectTest.java (hunk 5 above)
Comment (VRanga000) [inline apis/openstack-nova/src/test/java/org/jclouds/openstack/nova/v2_0/extensions/ServerGroupApiExpectTest.java]: Wildo

File: compute/src/test/java/org/jclouds/compute/internal/BaseComputeServiceLiveTest.java (hunk 7 above)
Comment (VRanga000) [inline compute/src/test/java/org/jclouds/compute/internal/BaseComputeServiceLiveTest.java]: No problem. Thanks for the review.

**Comment by andrewgaul** [comment]
@vapgar200 do we have a path forward on this pull request?

**Comment by gaul** [comment]
Closed due to inactivity.

**Pull request closed without merging**